### DISCUSSION TITLE
Finding the source of UVM error

### DISCUSSION THREAD

**Discussion by Sacusa** [NVIDIA/open-gpu-kernel-modules#691] [Q&A]
I am running a UVM CUDA application on a NUMA machine. I am trying to use `cudaMemAdvise_v2` to initialize the allocation to a DRAM-only NUMA node (i.e., no CPUs). Unfortunately, the call fails with the error: `NV_ERR_INVALID_ARGUMENT`. Looking at the driver code, my guess is that the problem is in `kernel_open/nvidia-uvm/uvm_policy:uvm_api_set_preferred_location`, but putting a `printk` inside the function does not print anything on either the console or `/var/log/kern.log` (the dmesg level is set to print everything).

I tried following the [previous advice on figuring the backtrace](https://github.com/NVIDIA/open-gpu-kernel-modules/discussions/337), but the first call to `nvidia_ioctl` has a `nv_printf` call that does not print anything either. I ensured `NVreg_RmMsg` is [set to print everything](https://github.com/NVIDIA/open-gpu-kernel-modules/discussions/197).

Any ideas what might be the problem here?

**Comment by mtijanic** [comment]
Hey there!

> but the first call to `nvidia_ioctl` has a `nv_printf` call that does not print anything either. I ensured `NVreg_RmMsg` is [set to print everything](https://github.com/NVIDIA/open-gpu-kernel-modules/discussions/197).

Can you try also setting `NVreg_ResmanDebugLevel=0` module param (or change [this line](https://github.com/NVIDIA/open-gpu-kernel-modules/blob/main/kernel-open/nvidia/os-interface.c#L802) to `= 0;`) and see if that gives you your `nv_printf()` output?

**Reply by Sacusa** [reply]
Thanks for the quick response, I was able to get the `ioctl` output! I noticed that executing `cudaMemAdvise_v2` does not print anything on the kernel logs. Do you have any pointers for that?

**Reply by mtijanic** [reply]
Not off the top of my head, sorry. How did you come to this conclusion, btw?
> Unfortunately, the call fails with the error: NV_ERR_INVALID_ARGUMENT

This error code is used internally by various driver components, but it should not be returned by the CUDA APIs. I expect it to fail with something like `CUDA_ERROR_INVALID_VALUE` instead. You're actually seeing the error value [0x1f](https://github.com/NVIDIA/open-gpu-kernel-modules/blob/main/src/common/sdk/nvidia/inc/nvstatuscodes.h#L60) being returned?

**Reply by Sacusa** [reply]
`CUDA_ERROR_INVALID_VALUE` (0x1) is indeed the error I receive, not `NV_ERR_INVALID_ARGUMENT`. I assumed that the the API error 0x1 mapped to driver error 0x1f based on my reading of the code.

**Reply by mtijanic** [reply]
In that case, I'm about 90% sure the issue is that the validation in the CUDA usermode library is failing your code before ever hitting UVM. Maybe post a minimal repro code?

**Reply by Sacusa** [reply]
Here you go: https://gist.github.com/Sacusa/d7faa2dbe8b2f6b3d23cbbb8907789b2

**Reply by Sacusa** [reply]
Am I correct in assuming that there is no way to debug the CUDA runtime code? Only the driver code seems to be open source.

**Reply by mtijanic** [reply]
Sorry, I'm not sure if there's a way to get any logs out of CUDA runtime. Maybe ask on the developer forums that are frequented by the CUDA devs?

I don't see anything immediately wrong with your code, but I hadn't had a chance to run it yet. Do you have HMM enabled?
https://developer.nvidia.com/blog/simplifying-gpu-application-development-with-heterogeneous-memory-management/#enabling_and_detecting_hmm

**Comment by Sur-Ring** [comment]
I don`t know how to config it, but I edited the source code and recompiled it to enable them. The first method is set the value of cur_debuglevel as 0xffffff8f. The second method is replace nv_printf with printk.
